# sequencing_summary

NOT finalised. This is just a draft subject to changes.

This sequencing_summary.txt generated by buttery-eel (when --seq_sum option is given) is close to ONT's sequencing summary as much as possible, but not identical.  The format specification is explained below.

sequencing_summary.txt is a tab-delimited text file with the first row being the header. The subsequent rows contain per-read information, where one row corresponds to one reads. The filed read_id is expected to have unique values. 

Note that columns have an arbitrary order. Use column header rather than column index when reading.

New additions are most likely added to the end unless it is some "important" value that makes more sense being next to some related value. However, it is requested not to rely on the column order, but rather use the header name. For extracting columns by header names in the order you require, we suggest using [csvkit](https://github.com/shenwei356/csvtk). For example, if you want to cut the columns with headers `read_id`, `parent_read_id` and `channel` (in order),  you may do: `csvtk -t cut sequencing_summary.txt -f read_id,parent_read_id,parent_read_id.

The columns are described below. Some of the descriptions are derived from the [ONT's document on sequencing summary](https://community.nanoporetech.com/docs/prepare/library_prep_protocols/Guppy-protocol/v/gpb_2003_v1_revax_14dec2018/input-and-output-files).

| Col | Type   | Name                     | Example value | Description |
| --- | ------ | ------------------------ | ----          | ----------- |
| 1   | string | filename_out             | reads.fastq    | name of the fastq/sam file this read was written to    |
| 2   | string | filename_slow5           | reads.blow5    | name of the slow5 file this read was read from     |
| 3   | string | parent_read_id           | <sub>f750bda8-0417-429f-b2ec-54c700543ed6 </sub>  | The original read_id if the read was split.  If not split, will be the same as the read_id column below.             |
| 4   | string | read_id                  | <sub>f750bda8-0417-429f-b2ec-54c700543ed6 </sub> | The unique read_id. This is a Universally unique identifier (UUID) version 4 and should be unique for any read from any device.  |
| 5   | string | run_id                   | <sub>dc60b20f5078b3546ded8... 10fb828b49c438fbd89322</sub> | The unique run ID which will be different for each run              |
| 6   | int    | channel                  | 10 | The channel number.  See Table 5 of [slow5 specification](https://hasindu2008.github.io/slow5specs/slow5-v1.0.0.pdf) for details. |
| 7   | int    | mux                      | 2 | The MUX setting for the channel when the read began. See Table 5 of [slow5 specification](https://hasindu2008.github.io/slow5specs/slow5-v1.0.0.pdf) for details    |
| 8   | int    | minknow_events           | 0 |  The number of events detected by MinKNOW. Zero if unknown, or if the value cannot be determined due to read-splitting  |
| 9   | int    | start_time               | 3034.378 | Start time of the read, in seconds since the beginning of the run. This is typically equal to the start_time in S/BLOW5 (which is in terms of number of samples), divided by the sampling_rate in S/BLOW5.   |
| 10  | int    | duration                 | 9.9695 | Time it took from start time to sequence read in seconds      |
| 11  | string | passes_filtering         | TRUE | TRUE/FALSE for passing the minimum qscore          |
| 12  | ?      | template_start           |   .   | Legacy value. |
| 13  | int    | num_events_template      |   .   | Legacy value. Number of events present in read.  |
| 14  | ?      | template_duration        | 16178 | Duration of the portion of the read that was sent to the basecaller after adapter trimming          |
| 15  | int    | sequence_length_template | 11.885449 | Number of bases in the output sequence, taking into account any sequence trimming.                        |
| 16  | float  | mean_qscore_template     | 3.239253 | The mean qscore of the read                        |
| 17  | float  | strand_score_template    | 106.933281 |  Legacy field - no longer populated reliably.    |
| 18  | float  | median_template          | 19.054329 | The median current of the read, in pA    |
| 19  | float  | mad_template             | 18.55447 | Median absolute difference for the read       |
| 21  | string | experiment_id            | PGXX22394 | Experiment ID if there is one present              |
| 22  | string | sample_id                | HG002 | The sample ID if there is one present              |
| 23  | string | end_reason               | signal_positive | The reason why the read ended (for the parent_read if the read was split). See  [slow5 specification](https://hasindu2008.github.io/slow5specs/slow5-v1.0.0.pdf) for details.                    |

If a certain value is missing or not applicable, it will be represented by a '.', for instance, the *passes_filtering* column when qscore filtering is not enabled.

# barcode_summary

When demultiplexing is enabled, buttery-ell will produce barcode_summary.txt which contains the following columns. The order of the columns is arbitrary, the same as for the sequenicng_summary.txt explained above. If a certain value is missing or not applicable, it will be represented by a '.'.

| Col | Type   | Name                          | Example value | Description |
| --- | ------ | ----------------------------- |----  | ----------- |
| 1   | string | parent_read_id                | <sub>88be501f-f700-4d9a-8341-13661d579fac</sub> 	| The original read_id if the read was split. If not split, it will be the same as the read_id column below.       |
| 2   | string | read_id                       | <sub>88be501f-f700-4d9a-8341-13661d579fac</sub>   | The unique read_id. This is a Universally unique identifier (UUID) version 4 and should be unique for any read from any device.                               |
| 3   | string | barcode_arrangement           | barcode02 								| The barcode name. Usually barcode<xx> or unclassifed                                |
| 4   | string | barcode_full_arrangement      | NB02_var2							    | The full name for the highest-scoring barcode match, including kit, variation, and direction	|
| 5   | string | barcode_kit                   | NB									    | Barcode kit the barcode comes from. Likely to be NB or RB for native and rapid               |
| 6   | string | barcode_variant               | var2								    | Which variant of the barcode was detected        |
| 7   | float  | barcode_score                 | 62.16666794 							| The alignment score of the barcode               |
| 8   | string | barcode_front_id              | NB02_FWD							    | The barcode name of the front detected barcode   |
| 9   | float  | barcode_front_score           | 62.16666794						    | the alignment score of the front barcode         |
| 10  | string | barcode_front_refseq          | <sub>ATTGCTAAGGTTAAACAGAC... GACTACAAACGGAATCGACAGCACC</sub> | The front barcode reference sequence             |
| 11  | string | barcode_front_foundseq        | <sub>ACTTGTAGGTGAAACAGACA... TCTAGCAACACTAAGAATGCCC</sub>	   | The sequence found to represent the front barcode|
| 12  | int    | barcode_front_foundseq_length | 42 | The length of the barcode_front_foundseq         |
| 13  | int    | barcode_front_begin_index     | 76 | The start position of the barcode_front_foundseq |
| 14  | string | barcode_rear_id               | NB02_REV | The barcode name of the rear detected barcode    |
| 15  | float  | barcode_rear_score            | 15.33333302 | the alignment score of the rear barcode          |
| 16  | string | barcode_rear_refseq           | <sub>AGGTGCTGTCGATTCCGTTT... GTAGTCGTCTGTTTAACCT</sub> | The rear barcode reference sequence              |
| 17  | string | barcode_rear_foundseq         | <sub>AGTGTGTCATGAGGGCACAA... ACAAATGTCTGTGTAAACAC</sub> | The sequence found to represent the rear barcode |
| 18  | int    | barcode_rear_foundseq_length  | 40 | The length of the barcode_rear_foundseq          |
| 19  | int    | barcode_rear_end_index        | 76 | The start position of the barcode_rear_foundseq  |


# FASTQ header tags

In FASTQ output generated by buttery-eel, following tags will be present as FASTQ header comments.

| Type   | Name                          | Example value | Description |
| ------ | ----------------------------- | --------------| ----------- |
| string | parent_read_id                |  <sub>f750bda8-0417-429f-b2ec-54c700543ed6</sub>             | The original read_id if the read was split. If not split, will be the same as the read_id.          |
| string | model_version_id              |  dna_r10.4.1_e8.2_400bps_fast-v4.1.0             | The basecalling model used           |
| int | mean_qscore                      |  11             | The mean qscore of the read           |
| string | barcode                       |  barcode10             | The barcode name. only if demultiplexing is enabled.            |

# SAM tags

In SAM output generated by buttery-eel, the following SAM tags will be present. 

Tags that are detailed in the SAM optional field specification [here](https://samtools.github.io/hts-specs/SAMtags.pdf): 
1. BC: barcode
2. ML and MM for methylation, only if --call_mods is specified

Non-standard tags: 

| Type   | Name               | Example value |Description |
| ------ | ------------------ | ------------- | ----------- |
| int (i) | qs                | 11 | the mean qscore of the read           |
| int (i) | ns                | 3196 | number of samples           |
| int (i) | ts                | 0 | number of trimmed samples           |
| string (Z) | pi                | <sub>f750bda8-0417-429f-b2ec-54c700543ed6</sub>  | The original read_id if the read was split. If not split, will be the same as the read_id.      |
| array (B) | mv                | <sub>mv:B:c:5,1,1,0,1,0,0,0,1,0,1...</sub>  | Move table. Only if --moves_out or --call_mods is specified. See [here](https://github.com/hasindu2008/squigualiser/blob/main/docs/move_table.md) for the explanation.  |.
| float (f) | sm | 104.84 | ? | 
| float (f) | sm | 18.2686 | ? | 
| string (Z) | sv    | <sub>quantile</sub>  | ?    |
